cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/MultipleOutputs.java

```diff
@@ -242,6 +245,47 @@ private static Job getNamedJob(JobContext context, String namedOutput) throws IO
     return job;
   }
 
+  /**
+   * Wraps RecordWriter to increment output counters.
+   */
+  @SuppressWarnings("unchecked")
+  private static class RecordWriterWithCounter extends RecordWriter {
+    private final RecordWriter writer;
+    private final String groupName;
+    private final String counterName;
+    private final TaskInputOutputContext context;
+
+    public RecordWriterWithCounter(RecordWriter writer, TaskInputOutputContext context) {
+      this.writer = writer;
+      this.context = context;
+      this.groupName = TaskCounter.class.getName();
+      this.counterName = getCounterName(context);
+    }
+
+    @SuppressWarnings("unchecked")
+    public void write(Object key, Object value) throws IOException, InterruptedException {
+      context.getCounter(groupName, counterName).increment(1);
+      writer.write(key, value);
+    }
+
+    public void close(TaskAttemptContext context) throws IOException, InterruptedException {
+      writer.close(context);
+    }
+
+    private String getCounterName(TaskInputOutputContext context) {
+      MapReduceMetrics.TaskType taskType = MapReduceMetrics.TaskType.from(context.getTaskAttemptID().getTaskType());
+      Preconditions.checkNotNull(taskType, "Illegal task type: " + context.getTaskAttemptID().getTaskType());
```

Comment:
There's a place where the code needs to know its null or not:
https://github.com/caskdata/cdap/blob/a7fb8ad579ca4fe14bbe63ac5fc27966e5b010cf/cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceTaskContextProvider.java#L165-165

I think its messy to do a try-catch for the IllegalArgumentException and translate it into `null` at that spot. What do you think?